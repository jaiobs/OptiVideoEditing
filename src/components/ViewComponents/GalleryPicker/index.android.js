/**
 * Video Editor screen
 * last edited: 16/10/2020
 */

import React, {Component} from 'react';
import {
  View,
  StyleSheet,
  TouchableOpacity,
  Text,
  Image,
  PermissionsAndroid,
  FlatList,
  Modal,
} from 'react-native';
import VideoPickerAndroid from '../../videopicker/VideoPickerAndroid';
import PhotoPickerAndroid from '../../photopicker/PhotoPickerAndroid';
import {Colors, Images} from '../../../res';
import utils from '../../../utils/utils';

export default class GalleryPicker extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activeTab: 0,
      isStoragePermissionGranted: false,
      videoFiles: [],
      imageFiles: [],
      pickerType: '',
    };
  }

  componentDidMount() {
    if (utils.isAndroid) {
      this.checkForStoragePermission();
    }
  }

  async checkForStoragePermission() {
    try {
      const granted = await PermissionsAndroid.request(
        PermissionsAndroid.PERMISSIONS.READ_EXTERNAL_STORAGE,
        {
          title: 'App Storage Permission',
          message: 'App needs access to your storage ',
          buttonNeutral: 'Ask Me Later',
          buttonNegative: 'Cancel',
          buttonPositive: 'OK',
        },
      );
      if (granted === PermissionsAndroid.RESULTS.GRANTED) {
        this.setState({isStoragePermissionGranted: true});
      }
    } catch (err) {
      console.warn(err);
    }
  }

  onDonePressed() {
    if (this.state.pickerType === 'video') {
      this.props.onDonePressed(this.state.videoFiles);
    } else if (this.state.imageFiles.length === 1) {
      var node = {
        node: {
          image: {
            uri: this.state.imageFiles.imagePath,
            item: this.state.imageFiles[0],
          },
        },
      };
      this.props.onPictureSelected(node);
    } else {
      this.props.onMultiplePictureSelected(this.state.imageFiles);
    }
  }

  onClosePressed() {
    this.setState({activeTab: 0, videoFiles: [], imageFiles: []});
    this.props.onClosePressed();
  }

  refreshCurrentPage() {
    if (this.state.activeTab === 0) {
      this.videoPicker.refresh();
    } else {
      this.imagePicker.refresh();
    }
  }

  onPressTab(selectedTab) {
    if (selectedTab === 0) {
      this.setState({activeTab: selectedTab, imageFiles: []});
    } else {
      this.setState({activeTab: selectedTab, videoFiles: []});
    }
  }
  _renderItemImage = ({item}) => {
    if (this.state.pickerType === 'video') {
      return (
        <View style={styles.bottomScrollImage}>
          <Image source={{uri: item.videoPath}} style={styles.imageSize} />
        </View>
      );
    } else {
      return (
        <View style={styles.bottomScrollImage}>
          <Image source={{uri: item.imagePath}} style={styles.imageSize} />
        </View>
      );
    }
  };

  render() {
    return (
      <Modal
        onShow={() => {
          this.refreshCurrentPage();
        }}
        animationType="slide"
        transparent={false}
        style={styles.container}
        onRequestClose={() => this.onClosePressed()}
        visible={this.props.visibility}>
        <View style={styles.headerContainer}>
          <View style={styles.headerActionContainer}>
            <TouchableOpacity
              style={styles.headerActionTouchable}
              onPress={() => this.onClosePressed()}>
              <Image source={Images.close_icon} style={styles.closeIcon} />
            </TouchableOpacity>
            {this.state.videoFiles.length >= 1 && (
              <TouchableOpacity
                style={styles.headerActionTouchable}
                onPress={() => this.onDonePressed()}>
                <Image source={Images.done_icon} style={styles.closeIcon} />
              </TouchableOpacity>
            )}
            {this.state.imageFiles.length >= 1 && (
              <TouchableOpacity
                style={styles.headerActionTouchable}
                onPress={() => this.onDonePressed()}>
                <Image source={Images.done_icon} style={styles.closeIcon} />
              </TouchableOpacity>
            )}
          </View>

          <View style={styles.tabContainer}>
            <View style={styles.leftHeaderContainer}>
              <TouchableOpacity onPress={() => this.onPressTab(0)}>
                <Text style={styles.tabTitle}>Videos</Text>
              </TouchableOpacity>
              {this.state.activeTab === 0 && (
                <View style={styles.activeUnderLiner} />
              )}
            </View>

            <View style={styles.rightHeaderContainer}>
              <TouchableOpacity onPress={() => this.onPressTab(1)}>
                <Text style={styles.tabTitle}>Photos</Text>
              </TouchableOpacity>
              {this.state.activeTab === 1 && (
                <View style={styles.activeUnderLiner} />
              )}
            </View>
          </View>

          <View
            style={[
              this.state.videoFiles.length >= 1 ? {paddingBottom: 80} : 'null',
            ]}>
            {this.state.activeTab === 0 && (
              <VideoPickerAndroid
                ref={(ref) => (this.videoPicker = ref)}
                width={this.props.width}
                rowItems={this.props.rowItems}
                orientation={this.props.orientation}
                onVideosSelected={(videoFiles) =>
                  this.setState({videoFiles, pickerType: 'video'})
                }
              />
            )}

            {this.state.activeTab === 1 && (
              <PhotoPickerAndroid
                ref={(ref) => (this.imagePicker = ref)}
                width={this.props.width}
                rowItems={this.props.rowItems}
                orientation={this.props.orientation}
                onItemSelected={(imageFiles) =>
                  this.setState({imageFiles, pickerType: 'image'})
                }
              />
            )}
          </View>

          <View style={styles.bottomImage}>
            <View style={styles.bottomImageView}>
              <View style={styles.imageView}>
                {this.state.pickerType === 'video' && (
                  <FlatList
                    horizontal
                    data={this.state.videoFiles}
                    renderItem={this._renderItemImage}
                    keyExtractor={(item) => item.id}
                  />
                )}
                {this.state.pickerType === 'image' && (
                  <FlatList
                    horizontal
                    data={this.state.imageFiles}
                    renderItem={this._renderItemImage}
                    keyExtractor={(item) => item.id}
                  />
                )}
              </View>
              <View style={styles.nextButton}>
                {this.state.videoFiles.length >= 1 && (
                  <TouchableOpacity
                    style={styles.headerActionTouchable}
                    onPress={() => this.onDonePressed()}>
                    <Image
                      source={Images.next_video_icon}
                      style={styles.closeIcon}
                    />
                  </TouchableOpacity>
                )}
                {this.state.imageFiles.length >= 1 && (
                  <TouchableOpacity
                    style={styles.headerActionTouchable}
                    onPress={() => this.onDonePressed()}>
                    <Image
                      source={Images.next_video_icon}
                      style={styles.closeIcon}
                    />
                  </TouchableOpacity>
                )}
              </View>
            </View>
          </View>
        </View>
      </Modal>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    backgroundColor: Colors.black,
    left: 0,
    top: 0,
    height: '100%',
    width: '100%',
    // zIndex: 10,
  },
  headerContainer: {
    flexDirection: 'column',
    backgroundColor: Colors.black,
  },
  leftHeaderContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  rightHeaderContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  bottomImage: {
    bottom: 0,
    justifyContent: 'flex-end',
    position: 'absolute',
    width: '100%',
    height: 100,
    marginBottom: 65,
  },
  bottomImageView: {
    flex: 1,
    flexDirection: 'row',
  },
  imageView: {flex: 1},
  nextButton: {
    width: 60,
    alignItems: 'center',
  },
  bottomScrollImage: {
    width: 70,
    paddingLeft: 10,
    paddingBottom: 40,
    height: '100%',
  },
  imageSize: {
    width: '100%',
    height: '100%',
  },
  closeIcon: {
    height: 24,
    width: 24,
  },
  activeUnderLiner: {
    top: 2,
    height: 6,
    backgroundColor: Colors.primaryAccent,
    width: 100,
  },
  tabContainer: {
    marginTop: 10,
    width: '100%',
    flexDirection: 'row',
  },
  tabTitle: {
    color: Colors.white,
    fontSize: 16,
    fontWeight: '700',
  },
  headerActionContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  headerActionTouchable: {
    marginTop: 4,
    marginBottom: 4,
    marginLeft: 10,
    marginRight: 10,
  },
});
