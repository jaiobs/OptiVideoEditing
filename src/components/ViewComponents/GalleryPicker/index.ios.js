import { Colors, Images } from "../../../res";
import {
  Image,
  Modal,
  PermissionsAndroid,
  SafeAreaView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from "react-native";
import React, { Component } from "react";

import PhotoPicker from "../../photopicker/PhotoPicker";
import VideoPicker from "../../videopicker/VideoPicker";

export default class GalleryPicker extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activeTab: 0,
      isStoragePermissionGranted: false
    };
  }

  async checkForStoragePermission() {
    try {
      const granted = await PermissionsAndroid.request(
        PermissionsAndroid.PERMISSIONS.READ_EXTERNAL_STORAGE,
        {
          title: "App Storage Permission",
          message: "App needs access to your storage ",
          buttonNeutral: "Ask Me Later",
          buttonNegative: "Cancel",
          buttonPositive: "OK"
        }
      );
      if (granted === PermissionsAndroid.RESULTS.GRANTED) {
        this.setState({ isStoragePermissionGranted: true });
      }
    } catch (err) {
      console.warn(err);
    }
  }

  render() {
    return (
      <Modal animationType="slide" supportedOrientations={['portrait', 'landscape']} transparent={false} onRequestClose={this.props.onClosePressed} visible={this.props.visibility}>
        <SafeAreaView style={styles.container}>
          <View style={{ flex: 1, flexDirection: 'column' }}>
            <View style={{ flex: 1,  maxHeight: 50 }} >
              <TouchableOpacity
                style={styles.closeBtnContainer}
                onPress={() => this.props.onClosePressed()}
              >
                <Image source={Images.close_icon} style={styles.closeIcon} />
              </TouchableOpacity>
            </View>

            <View style={{ flex: 1, maxHeight: 40, marginTop: this.props.orientation == "portrait" ? 0 : -40, width: this.props.orientation == "portrait" ? "100%" : "50%", alignSelf:"center"}} >
              <View style={styles.tabContainer}>
                <View style={styles.leftHeaderContainer}>
                  <TouchableOpacity onPress={() => this.setState({ activeTab: 0 })}>
                    <Text style={styles.tabTitle}>Videos</Text>
                  </TouchableOpacity>
                  {this.state.activeTab == 0 && (
                    <View style={styles.activeUnderLiner} />
                  )}
                </View>

                <View style={styles.rightHeaderContainer}>
                  <TouchableOpacity onPress={() => this.setState({ activeTab: 1 })}>
                    <Text style={styles.tabTitle}>Photos</Text>
                  </TouchableOpacity>
                  {this.state.activeTab == 1 && (
                    <View style={styles.activeUnderLiner} />
                  )}
                </View>
              </View>
            </View>
            <View style={{ flex: 1, marginTop: this.props.orientation == "portrait" ? 12 : 2}} >

              {this.state.activeTab == 0 && (
                <VideoPicker
                  width={this.props.width}
                  rowItems={this.props.rowItems}
                  orientation={this.props.orientation}
                  onItemSelected={item => {
                    this.props.onVideoSelected(item);
                  }}
                />
              )}

              {this.state.activeTab == 1 && (
                <PhotoPicker
                  width={this.props.width}
                  rowItems={this.props.rowItems}
                  orientation={this.props.orientation}
                  onItemSelected={item => { this.props.onPictureSelected(item) }}
                />
              )}


            </View>
          </View>

        </SafeAreaView>
      </Modal>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.black,
    left: 0,
    top: 0,
    height: "100%",
    width: "100%",
    flex: 1
  },
  headerContainer: {
    backgroundColor: Colors.black,
  },
  leftHeaderContainer: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center"
  },
  rightHeaderContainer: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center"
  },
  closeIcon: {
    height: 24,
    width: 24,
    resizeMode:"contain"
  },
  closeBtnContainer: {
    position: 'absolute',
    left: 15,
    top:10,
    paddingTop: 10,
    paddingRight: 10,
    paddingBottom: 10,
    paddingLeft: 10
  },
  activeUnderLiner: {
    top: 2,
    height: 6,
    backgroundColor: Colors.primaryAccent,
    width: 100
  },
  tabContainer: {
    marginTop: 10,
    width: "100%",
    flexDirection: "row"
  },
  tabTitle: {
    color: Colors.white,
    fontSize: 16,
    fontWeight: "700"
  }
});
