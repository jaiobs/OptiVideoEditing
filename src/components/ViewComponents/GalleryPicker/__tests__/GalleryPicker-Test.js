import React from 'react';
import renderer from 'react-test-renderer';
import GalleryPicker from '../index.android';
import {NativeModules} from 'react-native';

describe('GalleryPicker Tests', () => {
  beforeEach(() => {
    NativeModules.GalleryPickerModule = {
      getGalleryImages: (params) => {
        return Promise.resolve({
          page_info: {
            end_cursor: 0,
            has_next_page: 10,
          },
          edges: [
            {
              node: {
                image: {
                  url: 'test-image.png',
                },
              },
            },
          ],
        });
      },
      getGalleyVideos: (params) => {
        return Promise.resolve({
          page_info: {
            end_cursor: 0,
            has_next_page: 10,
          },
          edges: [
            {
              node: {
                video: {
                  url: 'test-video.mp4',
                },
              },
            },
          ],
        });
      },
    };
  });

  it('should render GalleryPicker correctly', () => {
    const tree = renderer.create(<GalleryPicker rowItems={3} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
