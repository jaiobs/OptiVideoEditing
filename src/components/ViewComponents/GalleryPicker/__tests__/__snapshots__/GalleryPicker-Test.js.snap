// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryPicker Tests should render GalleryPicker correctly 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  onShow={[Function]}
  style={
    Object {
      "backgroundColor": "#000",
      "height": "100%",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": "100%",
    }
  }
  transparent={false}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "#000",
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginBottom": 4,
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 4,
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/res/assets/images/cross.png",
            }
          }
          style={
            Object {
              "height": 24,
              "width": 24,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 16,
                "fontWeight": "700",
              }
            }
          >
            Videos
          </Text>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#f9089b",
              "height": 6,
              "top": 2,
              "width": 100,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 16,
                "fontWeight": "700",
              }
            }
          >
            Photos
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          "null",
        ]
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "height": "100%",
            "marginBottom": 12,
            "marginLeft": 6,
            "marginRight": 6,
            "marginTop": 12,
            "width": "100%",
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#f9089b"
          hidesWhenStopped={true}
          size="large"
          style={
            Object {
              "alignSef": "center",
            }
          }
        />
        <RCTScrollView
          nestedScrollEnabled={true}
          style={
            Object {
              "backgroundColor": "black",
              "flex": 1,
              "marginBottom": 200,
            }
          }
        >
          <View />
        </RCTScrollView>
      </View>
    </View>
    <View
      style={
        Object {
          "bottom": 0,
          "height": 100,
          "justifyContent": "flex-end",
          "marginBottom": 65,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "width": 60,
            }
          }
        />
      </View>
    </View>
  </View>
</Modal>
`;
